File: bwell-swift-ios/bwell-swift-ios/HealthSync/MetricCardView.swift
//
//  MetricCardView.swift
//  bwell-swift-ios
//
//  Shared card shape for the Metrics and Body Score grids - icon, title,
//  big value + unit, and an "Updated on <date>" footer. Ported in spirit
//  from ui-platform's mfe-devices DeviceVitalsCard (2-column grid of
//  rounded cards), simplified to this demo's needs (no reference-range
//  chip, no "just synced" highlight).
//

import SwiftUI

struct MetricCardView: View {
    let title: String
    let value: String?
    let unit: String?
    let footer: String?

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Image(systemName: HealthMetricIcon.symbolName(for: title))
                .font(.title2)
                .foregroundStyle(Color.accentColor)

            Text(title)
                .font(.subheadline).fontWeight(.semibold)
                .foregroundStyle(.primary)
                .lineLimit(2)
                .fixedSize(horizontal: false, vertical: true)

            if let value {
                VStack(alignment: .leading, spacing: 2) {
                    Text(value)
                        .font(.title2).fontWeight(.bold)
                        .foregroundStyle(.primary)
                    if let unit, !unit.isEmpty {
                        Text(unit)
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }
            }

            if let footer {
                Divider()
                Text(footer)
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
        }
        .padding()
        .frame(maxWidth: .infinity, minHeight: 150, alignment: .topLeading)
        .background(Color(.secondarySystemGroupedBackground))
        .clipShape(RoundedRectangle(cornerRadius: 16))
        .healthCardBorderStyle()
    }
}

/// Border + shadow + pointer-hover highlight shared by the Metrics grid
/// card and the Body Score row card - extracted since both used the exact
/// same three modifiers verbatim.
private struct HealthCardBorderStyle: ViewModifier {
    func body(content: Content) -> some View {
        content
            .overlay(
                RoundedRectangle(cornerRadius: 16)
                    .strokeBorder(Color(.separator), lineWidth: 1)
            )
            .shadow(color: .black.opacity(0.06), radius: 10, x: 0, y: 4)
            // Highlights the card under a pointer (trackpad/Magic Keyboard
            // on iPad) - the closest native equivalent to ui-platform's CSS
            // :hover/:focus box-shadow. Has no visible effect on iPhone
            // touch; CardPressButtonStyle below covers touch-driven feedback.
            .hoverEffect(.highlight)
    }
}

extension View {
    func healthCardBorderStyle() -> some View {
        modifier(HealthCardBorderStyle())
    }
}

/// Scales down and dims slightly while pressed - the touch-driven analog of
/// ui-platform's DeviceVitalsCard :hover/:focus state (stronger shadow,
/// darker border, hover background) for a device with no pointer.
struct CardPressButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.97 : 1)
            .opacity(configuration.isPressed ? 0.85 : 1)
            .animation(.easeOut(duration: 0.15), value: configuration.isPressed)
    }
}

/// Two-column grid, used identically by both the Metrics and Body Score tabs.
struct MetricCardGrid<Item: Identifiable, Content: View>: View {
    let items: [Item]
    @ViewBuilder let content: (Item) -> Content

    private let columns = [GridItem(.flexible(), spacing: 12), GridItem(.flexible(), spacing: 12)]

    var body: some View {
        LazyVGrid(columns: columns, spacing: 12) {
            ForEach(items) { item in
                content(item)
            }
        }
    }
}

File: bwell-kotlin-android/app/src/main/java/com/bwell/sampleapp/activities/ui/healthsync/MetricCardView.kt
package com.bwell.sampleapp.activities.ui.healthsync

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Divider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

/**
 * Shared card shape for the Metrics grid - icon, title, big value + unit,
 * and an "Updated on <date>" footer. Ported from Swift's MetricCardView
 * (itself ported from ui-platform's mfe-devices DeviceVitalsCard),
 * simplified to this demo's needs. Uses Compose's standard ripple
 * (via [clickable]) for press feedback rather than porting iOS's
 * scale+dim animation verbatim - the platform-idiomatic touch feedback for
 * the same underlying intent (visible pressed state).
 */
@Composable
fun MetricCardView(
    title: String,
    value: String?,
    unit: String?,
    footer: String?,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(16.dp)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .heightIn(min = 150.dp)
            .shadow(elevation = 4.dp, shape = shape, clip = false)
            .clip(shape)
            .background(Color.White)
            .border(1.dp, Color(0xFFE4E7EC), shape)
            .clickable(onClick = onClick)
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(
            HealthMetricIcon.iconFor(title),
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary,
            modifier = Modifier.size(28.dp),
        )
        Text(
            title,
            style = MaterialTheme.typography.titleSmall,
            fontWeight = FontWeight.SemiBold,
            maxLines = 2,
        )
        if (value != null) {
            Column {
                Text(value, style = MaterialTheme.typography.titleLarge, fontWeight = FontWeight.Bold)
                if (!unit.isNullOrEmpty()) {
                    Text(unit, style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
        }
        if (footer != null) {
            Divider()
            Text(footer, style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}
